Translated post react-v16.13.0 (#405)
reactjs/pt-BR.reactjs.org: content/blog/2020-02-26-react-v16.13.0.md

  - "blog/2020/03/02/react-v16.13.0.html"
---

> Hoje estamos lançando o React 16.13.0. Ele contém correções de bugs e novos avisos de descontinuação para ajudar a se preparar para uma futura versão principal.

Reviewer: Acho que seria legal seguirmos um padrão de tradução para o verbo `deprecate`, pois nesse artigo estamos usando descontinuar, depreciar e reprovar. Como vi que em outros artigos usamos depreciar, vou sugerir que todos sejam substituídos por esse.

```suggestion
Hoje estamos lançando o React 16.13.0. Ele contém correções de bugs e novos avisos de depreciação para ajudar a se preparar para uma futura versão principal.
```

</div> 
```

Você pode esperar essa `<div>` ter sempre um fundo vermelho, não importa o valor de `toggle`. No entanto, ao alternar o valor de`toggle` entre `true` e `false`, a cor de fundo começa como `red`, então alterna entre `transparent` e `blue`, [como você pode ver nessa demo](https://codesandbox.io/s/suspicious-sunset-g3jub).
  

> **O React agora detecta regras de estilo conflitantes e registra um aviso.** Para corrigir um problema, não misture versões shorthand e longhand da mesma propriedade CSS no `style` prop.

Reviewer 2: ```suggestion
**O React agora detecta regras de estilo conflitantes e registra um aviso.** Para corrigir um problema, não misture versões *shorthand* e *longhand* da mesma propriedade CSS na prop `style`.
```

> **O React agora detecta regras de estilo conflitantes e registra um aviso.** Para corrigir um problema, não misture versões shorthand e longhand da mesma propriedade CSS no `style` prop.

Reviewer 3: já foi ajustado em outro comentário :)

> ### Avisos para algumas refs de string obsoletas {#warnings-for-some-deprecated-string-refs}

Reviewer: ```suggestion
### Avisos para algumas string refs depreciadas {#warnings-for-some-deprecated-string-refs}
```

> [String Refs é uma legacy API](/docs/refs-and-the-dom.html#legacy-api-string-refs) que é desencorajada e será depreciada no futuro:

Reviewer: ```suggestion
[String Refs é uma API legada](/docs/refs-and-the-dom.html#legacy-api-string-refs) que é desencorajada e será depreciada no futuro:
```

```js
<Button ref="myRef" />
```

(Não confunda String Refs com refs em geral, o que **permanece totalmente suportado**.)

> No futuro, forneceremos um script automatizado (um "codemod") para migrar para longe das referências de string. No entanto, alguns casos raros não podem ser migrados automaticamente. Esta versão adiciona um novo aviso **apenas para esses casos** antes da reprovação.

Reviewer: ```suggestion
No futuro, forneceremos um script automatizado (um "codemod") para migrar para longe das referências de string. No entanto, alguns casos raros não podem ser migrados automaticamente. Esta versão adiciona um novo aviso **apenas para esses casos** antes da depreciação.
```

}
```

> Um código como esse geralmente indica bugs. (Você pode esperar que ref esteja disponível em `ClassParent`, mas em vez disso, é colocado em `ClassWithRenderProp`).

Reviewer: ```suggestion
Um código como esse geralmente indica bugs. (Você pode esperar que a ref esteja disponível em `ClassParent`, mas em vez disso, é colocado em `ClassWithRenderProp`).
```

}
```

> Nota
>

> > Para ver este aviso, você precisa ter o [babel-plugin-transform-react-jsx-self](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx-self) instalado nos seus plugins Babel. Ele deve  _apenas_ ser ativado no modo de desenvolvimento. 

Reviewer: ```suggestion
> Para ver este aviso, você precisa ter o [babel-plugin-transform-react-jsx-self](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx-self) instalado nos seus plugins Babel. Ele deve ser ativado _apenas_ no modo de desenvolvimento. 
```

> 
> Se você usa o Create React App ou tem a predefinição "react" com o Babel 7+, você já tem esse plug-in instalado por padrão.

> ### Descontinuação `React.createFactory` {#deprecating-reactcreatefactory}

Reviewer 2: ```suggestion
### Descontinuando `React.createFactory` {#deprecating-reactcreatefactory}
```

> ### Descontinuando `React.createFactory` {#deprecating-reactcreatefactory}

Reviewer: ```suggestion
### Depreciando `React.createFactory` {#deprecating-reactcreatefactory}
```

> [`React.createFactory`](/docs/react-api.html#createfactory) é um legado auxiliar para criar elementos React. Esta versão adiciona um aviso de descontinuação ao método. Ele será removido em uma versão principal futura.

Reviewer: ```suggestion
[`React.createFactory`](/docs/react-api.html#createfactory) é um helper legado para criar elementos React. Esta versão adiciona um aviso de depreciação ao método. Ele será removido em uma versão major futura.
```

Substitua os usos de `React.createFactory` com JSX regular. Como alternativa, você pode copiar e colar esse auxiliar de uma linha ou publicá-lo como uma biblioteca:

```jsx
let createFactory = type => React.createElement.bind(null, type);
```

Faz exatamente a mesma coisa.

> ### Descontinuação `ReactDOM.unstable_createPortal` in favor of `ReactDOM.createPortal` {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}

Reviewer 2: ```suggestion
### Descontinuando `ReactDOM.unstable_createPortal` em favor de `ReactDOM.createPortal` {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}
```

> ### Descontinuando `ReactDOM.unstable_createPortal` em favor de `ReactDOM.createPortal` {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}

Reviewer: ```suggestion
### Depreciando `ReactDOM.unstable_createPortal` em favor de `ReactDOM.createPortal` {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}
```

Quando o React 16 foi lançado,`createPortal` tornou-se uma API oficialmente suportada.

> No entanto, mantivemos `unstable_createPortal` como um alias suportado para manter as poucas bibliotecas que o adotaram funcionando. Agora estamos descontinuando o alias instável. Usar `createPortal` diretamente em vez de `unstable_createPortal`. Tem exatamente a mesma assinatura.

Reviewer: ```suggestion
No entanto, mantivemos `unstable_createPortal` como um alias suportado para manter as poucas bibliotecas que o adotaram funcionando. Agora estamos depreciando o alias instável. Use `createPortal` diretamente em vez de `unstable_createPortal`. Tem exatamente a mesma assinatura.
```

> ## Outras melhorias {#other-improvements}

Reviewer 2: ```suggestion
## Outras Melhorias {#other-improvements}
```

![Uma captura de tela do aviso do console, simplesmente indicando a natureza da incompatibilidade de hidratação: "Aviso: o HTML do servidor esperado deve conter uma div correspondente em div."](../images/blog/react-v16.13.0/hydration-warning-before.png)

Embora esteja apontando um erro no código, não está claro onde o erro existe e o que fazer em seguida. Esta versão adiciona uma pilha de componentes a este aviso, o que faz com que fique assim:

![Uma captura de tela do aviso do console, informando a natureza da incompatibilidade de hidratação, mas também incluindo uma pilha de componentes: "Aviso: o HTML do servidor esperado deve conter uma div correspondente em div, em div(no pages/index.js:4)..."](../images/blog/react-v16.13.0/hydration-warning-after.png)

Isso deixa claro onde está o problema e permite localizar e corrigir o bug mais rapidamente.

### Correções de bugs importantes {#notable-bugfixes}

Esta versão contém algumas outras melhorias importantes:

> - No Modo de Desenvolvimento Rigoroso, o React chama os métodos do ciclo de vida duas vezes para eliminar possíveis efeitos colaterais indesejados. Esta versão adiciona esse comportamento a `shouldComponentUpdate`. Isso não deve afetar a maioria dos códigos, a menos que você tenha efeitos colaterais em `shouldComponentUpdate`. Para corrigir isso, mova o código com efeitos colaterais para `componentDidUpdate`.

Reviewer: ```suggestion
- No Modo Estrito de Desenvolvimento, o React chama os métodos do ciclo de vida duas vezes para eliminar possíveis efeitos colaterais indesejados. Esta versão adiciona esse comportamento a `shouldComponentUpdate`. Isso não deve afetar a maioria dos códigos, a menos que você tenha efeitos colaterais em `shouldComponentUpdate`. Para corrigir isso, mova o código com efeitos colaterais para `componentDidUpdate`.
```

> - No Modo de Desenvolvimento Rigoroso, os avisos para o uso da API de contexto herdado não incluíam a pilha do componente que acionou o aviso. Esta versão adiciona a stack ausente ao aviso.

Reviewer: ```suggestion
- No Modo Estrito de Desenvolvimento, os avisos para o uso da API de contexto legada não incluíam a pilha do componente que acionou o aviso. Esta versão adiciona a pilha ausente ao aviso.
```

> - `onMouseEnter` agora não é ativado quando desativado `<button>` elementos.

Reviewer 2: ```suggestion
- `onMouseEnter` agora não é disparado em elementos `<button>` desabilitados.
```

> - O ReactDOM estava ausente `version` exportar desde que publicamos a v16. Esta versão adiciona de volta. Não recomendamos o uso na lógica do aplicativo, mas é útil ao depurar problemas com incompatibilidade / várias versões do ReactDOM na mesma página.

Reviewer 2: ```suggestion
- O ReactDOM tinha uma exportação de `version` ausente desde que publicamos a v16. Esta versão adiciona de volta. Não recomendamos o uso na lógica do aplicativo, mas é útil ao depurar problemas com incompatibilidade / várias versões do ReactDOM na mesma página.
```

> Agradecemos a todos os colaboradores que ajudaram a resolver e solucionar esses e outros problemas. Você pode encontrar o changelog completo [below](#changelog).

Reviewer 2: ```suggestion
Agradecemos a todos os colaboradores que ajudaram a resolver e solucionar esses e outros problemas. Você pode encontrar o changelog completo [abaixo](#changelog).
```

## Instalação {#installation}

### React {#react}

O React v16.13.0 está disponível no registro npm.

Para instalar o React 16 com o Yarn, execute:

```bash
yarn add react@^16.13.0 react-dom@^16.13.0
```

Para instalar o React 16 com npm, execute:

```bash
npm install --save react@^16.13.0 react-dom@^16.13.0
```

Também fornecemos compilações UMD do React por meio de uma CDN:

```html
<script crossorigin src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
<script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
```

> Consulte a documentação para[detailed installation instructions](/docs/installation.html).

Reviewer 2: ```suggestion
Consulte a documentação para [instruções detalhadas de instalação](/docs/installation.html).
```

## Changelog {#changelog}

### React {#react}

> - Avisa quando uma string ref é usada de uma maneira que não seja passível de um futuro código ([@lunaruan](https://github.com/lunaruan) in [#17864](https://github.com/facebook/react/pull/17864))

Reviewer: ```suggestion
- Avisa quando uma string ref é usada de uma maneira que não seja compatível com um futuro codemod ([@lunaruan](https://github.com/lunaruan) in [#17864](https://github.com/facebook/react/pull/17864))
```

> - Descontinua `React.createFactory()` ([@trueadm](https://github.com/trueadm) in [#17878](https://github.com/facebook/react/pull/17878))

Reviewer: ```suggestion
- Deprecia `React.createFactory()` ([@trueadm](https://github.com/trueadm) in [#17878](https://github.com/facebook/react/pull/17878))
```

### React DOM {#react-dom}

> - Avisar quando houver alterações `style` pode causar uma colisão inesperada ([@sophiebits](https://github.com/sophiebits) in [#14181](https://github.com/facebook/react/pull/14181), [#18002](https://github.com/facebook/react/pull/18002))

Reviewer 2: ```suggestion
- Avisar quando houver alterações `style` pode causar uma colisão inesperada ([@sophiebits](https://github.com/sophiebits) em [#14181](https://github.com/facebook/react/pull/14181), [#18002](https://github.com/facebook/react/pull/18002))
```

> - Avisa quando houver alterações `style` pode causar uma colisão inesperada ([@sophiebits](https://github.com/sophiebits) em [#14181](https://github.com/facebook/react/pull/14181), [#18002](https://github.com/facebook/react/pull/18002))

Reviewer: ```suggestion
- Avisa quando houver alterações `style` podem causar uma colisão inesperada ([@sophiebits](https://github.com/sophiebits) em [#14181](https://github.com/facebook/react/pull/14181), [#18002](https://github.com/facebook/react/pull/18002))
```

> - Avisar quando um componente de função é atualizado durante a fase de renderização de outro componente ([@acdlite](<(https://github.com/acdlite)>) in [#17099](https://github.com/facebook/react/pull/17099))

Reviewer 2: ```suggestion
- Avisar quando um componente de função é atualizado durante a fase de renderização de outro componente ([@acdlite](<(https://github.com/acdlite)>) em [#17099](https://github.com/facebook/react/pull/17099))
```

- Depreciar `unstable_createPortal` ([@trueadm](https://github.com/trueadm) em [#17880](https://github.com/facebook/react/pull/17880))

> - Consertar `onMouseEnter` sendo acionado nos botões desativados ([@AlfredoGJ](https://github.com/AlfredoGJ) in [#17675](https://github.com/facebook/react/pull/17675))

Reviewer 2: ```suggestion
- Consertar `onMouseEnter` sendo acionado nos botões desativados ([@AlfredoGJ](https://github.com/AlfredoGJ) em [#17675](https://github.com/facebook/react/pull/17675))
```

- Chama `shouldComponentUpdate` duas vezes ao desenvolver em `StrictMode` ([@bvaughn](https://github.com/bvaughn) em [#17942](https://github.com/facebook/react/pull/17942))
- Adiciona propriedade `version` para ReactDOM ([@ealush](https://github.com/ealush) em [#15780](https://github.com/facebook/react/pull/15780))

> - Não chame `toString()` em `dangerouslySetInnerHTML` ([@sebmarkbage](https://github.com/sebmarkbage) em [#17773](https://github.com/facebook/react/pull/17773))

Reviewer: ```suggestion
- Não chamar `toString()` em `dangerouslySetInnerHTML` ([@sebmarkbage](https://github.com/sebmarkbage) em [#17773](https://github.com/facebook/react/pull/17773))
```

- Mostra pilhas de componentes em mais avisos ([@gaearon](https://github.com/gaearon) em [#17922](https://github.com/facebook/react/pull/17922), [#17586](https://github.com/facebook/react/pull/17586))

> ### Modo Simultâneo (Experimental) {#concurrent-mode-experimental}

Reviewer: ```suggestion
### Modo Concorrente (Experimental) {#concurrent-mode-experimental}
```

- Avisa sobre usos problemáticos de `ReactDOM.createRoot()` ([@trueadm](https://github.com/trueadm) em [#17937](https://github.com/facebook/react/pull/17937))
- Retira `ReactDOM.createRoot()` parâmetros de retorno de chamada e avisos adicionais sobre o uso ([@bvaughn](https://github.com/bvaughn) em [#17916](https://github.com/facebook/react/pull/17916))
- Não agrupe o trabalho ocioso / fora da tela com outro trabalho ([@sebmarkbage](https://github.com/sebmarkbage) em [#17456](https://github.com/facebook/react/pull/17456))

> - Ajusta `SuspenseList` CPU bound heuristic ([@sebmarkbage](https://github.com/sebmarkbage) em [#17455](https://github.com/facebook/react/pull/17455))

Reviewer: ```suggestion
- Ajusta heurística de limite de processamento do `SuspenseList` ([@sebmarkbage](https://github.com/sebmarkbage) em [#17455](https://github.com/facebook/react/pull/17455))
```

- Adiciona prioridades de plugins de eventos ausentes ([@trueadm](https://github.com/trueadm) em [#17914](https://github.com/facebook/react/pull/17914))

> - Conserta `isPending` somente sendo verdade ao fazer a transição de dentro de um evento de entrada ([@acdlite](https://github.com/acdlite) em [#17382](https://github.com/facebook/react/pull/17382))

Reviewer: ```suggestion
- Conserta `isPending` somente sendo verdadeiro ao fazer a transição de dentro de um evento de input ([@acdlite](https://github.com/acdlite) em [#17382](https://github.com/facebook/react/pull/17382))
```

> - Conserta componente `React.memo` que descartam atualizações quando interrompidos por uma atualização de prioridade mais alta ([@acdlite](<(https://github.com/acdlite)>) em [#18091](https://github.com/facebook/react/pull/18091))

Reviewer: ```suggestion
- Conserta componentes `React.memo` que descartam atualizações quando interrompidos por uma atualização de prioridade mais alta ([@acdlite](<(https://github.com/acdlite)>) em [#18091](https://github.com/facebook/react/pull/18091))
```
